core/src/Cabana_Parallel.hpp

```diff
@@ -1059,6 +1061,217 @@ for_each_neighbor( const IndexType i, const TeamMemberType team,
         } );
 }
 
+//---------------------------------------------------------------------------//
+// Linked Cell Parallel For
+//---------------------------------------------------------------------------//
+
+//---------------------------------------------------------------------------//
+/*!
+  \brief Execute functor in parallel according to the execution policy over
+  particles with thread-local serial loops over linked cell list bins and
+  particle first neighbors in those bins.
+
+  \tparam FunctorType The functor type to execute.
+  \tparam NeighborListType The neighbor list type.
+  \tparam ExecParams The Kokkos range policy parameters.
+
+  \param exec_policy The policy over which to execute the functor.
+  \param functor The functor to execute in parallel
+  \param list The linked cell list over which to execute neighbor operations.
+  \param cell_stencil Stencil of cell neighbors to iterate over.
+  \param FirstNeighborsTag Tag indicating operations over particle first
+  neighbors.
+  \param SerialOpTag Tag indicating a serial loop strategy over neighbors.
+  \param str Optional name for the functor. Will be forwarded if non-empty to
+  the Kokkos::parallel_for called by this code and can be used for
+  identification and profiling purposes.
+
+  A "functor" is a class containing the function to execute in parallel, data
+  needed for that execution, and an optional \c execution_space typedef.  Here
+  is an example functor for neighbor parallel_for:
+
+  \code
+  class FunctorType {
+  public:
+  typedef  ...  execution_space ;
+  void operator() ( const int particle_index, const int neighbor_index ) const ;
+  };
+  \endcode
+
+  In the above example, \c Index is a Cabana index to a given AoSoA element
+  for a particle and its neighbor. Its <tt>operator()</tt> method defines the
+  operation to parallelize, over the range of indices
+  <tt>idx=[begin,end]</tt>.  This compares to a single iteration \c idx of a
+  \c for loop.
+*/
+template <class FunctorType, class LinkedCellType, class StencilType,
+          class... ExecParameters>
+inline void linkedcell_parallel_for(
+    const Kokkos::RangePolicy<ExecParameters...>& exec_policy,
+    const FunctorType& functor, const LinkedCellType& list,
+    const StencilType& cell_stencil, const FirstNeighborsTag, const SerialOpTag,
+    const std::string& str = "" )
+{
+    using work_tag = typename Kokkos::RangePolicy<ExecParameters...>::work_tag;
+
+    using execution_space =
+        typename Kokkos::RangePolicy<ExecParameters...>::execution_space;
+
+    using index_type =
+        typename Kokkos::RangePolicy<ExecParameters...>::index_type;
+
+    using memory_space = typename LinkedCellType::device_type::memory_space;
+
+    NeighborDiscriminator<FullNeighborTag> discriminator;
```

Comment:
Does the half list ever make sense in this context or always just full?

Reply:
I think based on the way you wrote the half discriminator it could work, but I would prefer to leave that until last